sec_certs/sample/common_criteria.py

```diff
@@ -684,6 +696,41 @@ def compute_heuristics_cert_id(self, all_cert_ids: Set[str]):
         self.heuristics.cert_id = self.pdf_data.cert_id
         self.normalize_cert_id(all_cert_ids)
 
+    def _get_dependency_cves(self, dset: "CCDataset", dependency_type: DependencyCVEType) -> Optional[Set[str]]:
+        dependency_type_dict = {
+            "direct": self.heuristics.report_references.directly_referenced_by,
+            "indirect": self.heuristics.report_references.indirectly_referenced_by,
+        }
+
+        references = dependency_type_dict[dependency_type.value]
+
+        if not references:
+            return None
+
+        vulnerabilities = set()
+        dataset_cert_id_occurrences = dset.get_dataset_cert_ids_occurrences()
+
+        for cert_id in references:
+            cert_id_occurrence = dataset_cert_id_occurrences.get(cert_id, None)
+
+            # Not in dataset or more than 2 occurrences in dataset -> skip
+            if cert_id_occurrence is None or cert_id_occurrence >= 2:
+                continue
+
+            for cert_obj in dset:
+                if cert_obj.heuristics.cert_id == cert_id:
+                    if cert_obj.heuristics.related_cves:
+                        vulnerabilities.update(cert_obj.heuristics.related_cves)
+
+        if not vulnerabilities:
+            return None
+
+        return vulnerabilities
+
+    def find_certificate_dependency_vulnerabilities(self, dset: "CCDataset"):
+        self.heuristics.direct_vulnerabilities = self._get_dependency_cves(dset, DependencyCVEType.DIRECT)
+        self.heuristics.indirect_vulnerabilities = self._get_dependency_cves(dset, DependencyCVEType.INDIRECT)
```

Comment:
**Thought: It's slightly weird that a `CommonCriteriaCert` needs `CCDataset` class to compute this.

Can we maybe delete this method and perform the respective calls directly from the `CCDataset` object? I'm improvizing, but `CCDataset` class could have function roughly like

```Python
for cert in self:
    cert._get_dependency_cves(self, DependencyCVEType.DIRECT)
```

the `_get_dependency_cves` would need to be exposed as public though.

Reply:
Naah, that doesn't help, it just moves the problem one step below, as `_get_dependency_cves()` should ideally not need the `CCDataset` variable as well...

Well, the most clean option would be to implement something in `model` module, similarly to the finder of references. That class would stand on top of both `CCDataset` and the certificates...

I understand that this would, however, require substantial refactoring. Leaving it to your consideration.